src/compact.jl

```diff
@@ -127,16 +125,19 @@ macro compact(fex, _kwexs...)
   end)
 end
 
-function addprefix!(ex::Expr, self, vars)
-  for i = 1:length(ex.args)
-    if ex.args[i] in vars
-      ex.args[i] = :($self.$(ex.args[i]))
-    else
-      addprefix!(ex.args[i], self, vars)
+function supportself(fex::Expr, vars)
+  @gensym self
+  @gensym curried_fex
+  # To avoid having to manipulate fex's arguments and body explicitly, we form a curried function first
+  # that wraps the full fex expression, and then uncurry it programatically rather than syntactically.
+  let_exprs = map(var -> :($var = $self.$var), vars)
+  return quote
+    $curried_fex = ($self) -> let $(let_exprs...) 
+        $fex
     end
+    ($self, args...; kwargs...) -> begin $curried_fex($self)(args...; kwargs...) end
```

Comment:
Trying to think of examples where weird things might happen, are we worse off in cases like this:
```julia
@compact(T = Float32, f = sin) do x
    f(x) / log(T(2))
end
```
where the anon function may not specialise on concrete subtypes of Type, Function? I struggle to see this mattering much. 

Are there any other weird issues to beware of? Might be worth timing Zygote to see if these nested functions cause it to behave worse.

Reply:
Do you recommend caring about Zygote's performance on small inputs? Zygote indeed gets a bit troubled by the currying machinery and slows down on e.g. arrays of length 3.

Reply:
The conclusion after benchmarking is that on problems where the gradient calls takes <= 50 microseconds, the curried function approach, as opposed to directly inlining the function body and syntactically currying, can incur significant overhead (e.g. gradients that took 50 microseconds can take 100 microsends with this PR).

There are no type instability issues, and for gradients taking around 1ms the overhead is not noticeable. Essentially Zygote just has strong opinions between things like `(x -> sum(w .* x))(args...)` and `sum(w .* args[1])`.

Unless you think otherwise (@MilesCranmer would your use cases care about overhead for very small networks?) I'm hoping to stick to the current approach: the idea of preserving the user-provided lambda, and not touching its syntax at least as far as functionality (i.e. not printing) goes, is very appealing, and the slowdown is technically a Zygote issue

Reply:
Thanks for looking. I'm mostly inclined to agree.

Reply:
Sounds good to me. I think safety > speed when the speedup is only 50 us.